x-pack/filebeat/input/gcs/job.go

```diff
@@ -121,57 +117,209 @@ func (j *job) Timestamp() time.Time {
 	return j.object.Updated
 }
 
-func (j *job) extractData(ctx context.Context) (*bytes.Buffer, error) {
+func (j *job) processAndPublishData(ctx context.Context, id string) error {
 	var err error
+	var offset int64
 	ctxWithTimeout, cancel := context.WithTimeout(ctx, j.src.BucketTimeOut)
 	defer cancel()
 	obj := j.bucket.Object(j.object.Name)
-	reader, err := obj.NewReader(ctxWithTimeout)
+	// if object is compressed or object root element is an array, then we cannot use an
+	// offset to read as it will produce an erroneous data stream.
+	if !j.isCompressed && !j.isRootArray {
+		offset = j.offset
+	}
+	reader, err := obj.NewRangeReader(ctxWithTimeout, offset, -1)
 	if err != nil {
-		return nil, fmt.Errorf("failed to read data from object with error: %w", err)
+		return fmt.Errorf("failed to open reader for object: %s, with error: %w", j.object.Name, err)
 	}
 	defer func() {
 		err = reader.Close()
 		if err != nil {
-			j.log.Errorf("failed to close object reader with error: %w", err)
+			j.log.Errorf("failed to close reader for object: %s, with error: %w", j.object.Name, err)
 		}
 	}()
 
-	data := &bytes.Buffer{}
+	updatedReader, err := j.addGzipDecoderIfNeeded(reader)
+	if err != nil {
+		return fmt.Errorf("failed to add gzip decoder to object: %s, with error: %w", j.object.Name, err)
+	}
+	err = j.readJsonAndPublish(ctx, updatedReader, id)
+	if err != nil {
+		return fmt.Errorf("failed to read data from object: %s, with error: %w", j.object.Name, err)
+	}
+
+	return err
+}
 
-	_, err = data.ReadFrom(reader)
+// addGzipDecoderIfNeeded determines whether the given stream of bytes (encapsulated in a buffered reader)
+// represents gzipped content or not and adds gzipped decoder if needed. A buffered reader is used
+// so the function can peek into the byte  stream without consuming it. This makes it convenient for
+// code executed after this function call to consume the stream if it wants.
+func (j *job) addGzipDecoderIfNeeded(body io.Reader) (io.Reader, error) {
+	bufReader := bufio.NewReader(body)
```

Comment:
Rather than wrapping here, maybe make this function take a `readPeeker` interface type that satisfies `io.Reader` and `Peek(int) ([]byte, error)` and then hoist the wrapping. Same for `evaluateJSON`.

Reply:
I'm not understanding what's the added advantage if we do this, since there is no native readPeeker interface type and will have to create a custom one.

Reply:
You can make a take a `*bufio.Reader`, same thing really. The idea is to hoist the repeated calls to `bufio.NewReader` out of the functions.

Reply:
ok got it.

Reply:
@efd6 have updated it as per suggestions but had to put in some extra code for the scenario of a gzipped stream, since for gzipped streams the underlying reader is of type *gzip.Reader, so have to again convert it to a *bufio.Reader before passing it to evaluateJSON().